### PR TITLE
KAFKA-13759: Disable idempotence by default in producers instantiated by Connect

### PR DESCRIPTION
**Pull request by kkonstantine**
With AK 3.0, idempotence was enabled by default in Kafka producers. 

However, if idempotence is enabled, Connect won't be able to communicate via its producers with Kafka brokers older than version 0.11. Perhaps more importantly, for brokers older than version 2.8 the IDEMPOTENT_WRITE ACL is required to be granted to the principal of the Connect worker. 

Therefore this commit disables producer idempotence by default to all the producers instantiated by Connect. Users can still choose to enable producer idempotence by explicitly setting the right worker and/or connector properties. 

The changes were tested via existing unit, integration and system tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 1]
```diff
@@ -648,6 +648,14 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
         // These settings will execute infinite retries on retriable exceptions. They *may* be overridden via configs passed to the worker,
         // but this may compromise the delivery guarantees of Kafka Connect.
         producerProps.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, Long.toString(Long.MAX_VALUE));
+        // By default, producers that are instantiated and used by Connect have idempotency disabled even after idempotency became
+        // default for Kafka producers. This is chosen to avoid breaking changes when Connect contacts Kafka brokers that do not support
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Maybe rephrase to be more positive? Like:
```suggestion
        // default for Kafka producers. This ensures Connect continues to work with many Kafka broker versions, including older brokers that do not support
```
If you make a change like this, please change the other places to be consistent.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 2]
```diff
@@ -648,6 +648,14 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
...
+        // idempotent producers or require explicit steps to enable them (e.g. adding the IDEMPOTENT_WRITE ACL to brokers older than 2.8).
+        // These settings might change when https://cwiki.apache.org/confluence/display/KAFKA/KIP-318%3A+Make+Kafka+Connect+Source+idempotent
+        // gets approved and scheduled for release.
+        producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "false");
+        producerProps.put(ProducerConfig.ACKS_CONFIG, "all");
+        producerProps.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, "1");
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Didn't #11932 add these two lines back in? If so, the only change should be the addition of:
```
producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "false");
```
Maybe this PR just needs to merge the latest from `trunk`.

**Review by rhauch — COMMENTED** [review]
Thanks, @kkonstantine. A pretty minor wording suggestion below.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 1 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Thanks for the suggestion. Added it in all places

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 2 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: I had included the other commit on purpose given that both PRs went out at the same time in order to present the same final outcome. I have rebased in the meantime after merging https://github.com/apache/kafka/pull/11932

**Comment by kkonstantine** [comment]
Thanks for the first review @rhauch
I have incorporated your feedback and have added a note in the upgrade notes. I'd appreciate a review on that text as well.

**Comment by kkonstantine** [comment]
Tests passed on the latest run with a couple of failures that don't seem relevant. Yet, because one failure was on `SourceConnectorsIntegrationTest` I'm rerunning these tests.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 3]
```diff
@@ -648,6 +648,12 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
         // These settings will execute infinite retries on retriable exceptions. They *may* be overridden via configs passed to the worker,
         // but this may compromise the delivery guarantees of Kafka Connect.
         producerProps.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, Long.toString(Long.MAX_VALUE));
+        // By default, producers that are instantiated and used by Connect have idempotence disabled even after idempotence became
+        // default for Kafka producers. This ensures Connect continues to work with many Kafka broker versions, including older brokers that do not support
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Minor nit, but I think it's worth making it super clear that _Connect_ is disabling the idempotent behavior of all it's producers, versus somehow the _producers_ doing something to disable idempotent behavior.
```suggestion
        // By default, Connect disables idempotent behavior for all producers, even though idempotence became
        // default for Kafka producers. This is to ensure Connect continues to work with many Kafka broker versions, including older brokers that do not support
```

File: docs/upgrade.html [hunk 4]
```diff
@@ -25,6 +25,10 @@ <h5><a id="upgrade_320_notable" href="#upgrade_320_notable">Notable changes in 3
             which meant that idempotence remained disabled unless the user had explicitly set <code>enable.idempotence</code> to true
             (See <a href="https://issues.apache.org/jira/browse/KAFKA-13598">KAFKA-13598</a>for more details).
             This issue was fixed and the default is properly applied in 3.0.1, 3.1.1, and 3.2.0.</li>
+        <li>A notable exception is Connect that by default currently disables idempotence for its
+            producers in order to support connecting to older versions of the Kafka broker without
+            explicit changes. Enabling idempotence is still possible via configuration and this
+            default might change in a major version upgrade in the future.</li>
```
Comment (rhauch) [inline docs/upgrade.html]: I also think it's worth clarifying the same thing here. Maybe:
```suggestion
        <li>A notable exception is Connect that by default disables idempotent behavior for all of its
            producers in order to uniformly support using a wide range of Kafka broker versions. 
            Users can change this behavior to enable idempotence for some or all producers
            via Connect worker and/or connector configuration. Connect may enable idempotent producers
            by default in a future major release.</li>
```
WDYT?

**Review by rhauch — COMMENTED** [review]
Just a few minor suggestions to wordsmith the comments and the Upgrade user documentation. Otherwise, LGTM.

**Comment by kkonstantine** [comment]
Thanks @rhauch. I've incorporated your suggestions in all places. The PR is ready for another pass

**Review by rhauch — APPROVED** [review]
LGTM. Thanks, @kkonstantine!

**Comment by kkonstantine** [comment]
Thanks @rhauch. The fix has now been merged to trunk and cherry picked to 3.2, 3.1 and 3.0 with the appropriate adjustments to the upgrade notes. cc @cadonna @tombentley
